### PR TITLE
FEATURE: Support search by property & exact value in NodeDataRepository

### PR DESCRIPTION
**Pull request by daniellienert**
Currently it is only possible to search through the properties by
giving a string that matches for any key or value found in the
jsonified properties field.

With this change, the term can also be an array to match exactly on a
given key / value combination.
The search term could be given as `['key' => 'value']`.

NEOS-1460 #close

### CODE REVIEW

**Comment by hlubek** [comment]
Hi Daniel,

pretty nice feature! I'd really like to have a functional test for this because we had a lot of trouble with database (text) search and MySQL / PostgreSQL / Sqlite compatibility. It would also be much easier to cover different cases in the future.

One other thing: I think we should convert the given term the same way we encode the JSON value in a node. That way we could search for a Node reference or non-string values too.

**Comment by daniellienert** [comment]
@hlubek - I added a test for the simple search in the property field and the search for a specific named property.

**Comment by skurfuerst** [comment]
@hlubek ping ;-)

**Comment by kdambekalns** [comment]
TL;DR: :-1: 

Ok, first of all – what about closing this PR and creating a new one? It seems the Travis reporting is broken for "old" PRs like this one…

The change looks good, and works most of the time on MySQL and (probably) never on PostgreSQL. Why? Because the search query does not specify an ordering so the `findNodeByPropertySearch` test (relying on the order of returned nodes) is not really reliable.

The question now is: Is the order of returned node important and should be made consistent? Or should the test be adjusted to check for the expected results even when the order changes?

**Comment by kdambekalns** [comment]
Here is an "example failure" on MySQL: https://travis-ci.org/kdambekalns/neos-development-collection/jobs/89391932

**Comment by aertmann** [comment]
Closed/reopened to fix test run

@kdambekalns: The ordering issue is unrelated to the actual change though, but made visible because of the test..

Adding order is potentially a breaking change, although a improvement. Also what ordering even makes sense in this case? A "semi random" one using the identifiers or the creation date or last modified date?

Because of that I'd suggest to make the test order independent and tackle the ordering in a separate change.

**Comment by kitsunet** [comment]
Thinking how we use this, order related to paths could make sense maybe

**Comment by aertmann** [comment]
@kitsunet: been down that road, it's a dead end.. ordering in SQL by path and sort index correctly is a nightmare

**Comment by kitsunet** [comment]
right, both won't work, I am aware. I just thought path plain and simple... Although change date could be an interesting choice thinking about it. DESC then.

**Comment by kitsunet** [comment]
Can we fix the last comments and get this in?

**Comment by kdambekalns** [comment]
Sure, so we'd go the "make the test check independent of order"?

**Comment by kitsunet** [comment]
sounds good to me!

**Comment by daniellienert** [comment]
Uh! *excited* ;) I will adjust the test then.

**Comment by daniellienert** [comment]
Mhm, seems that the unit /functional tests are not triggered atm? Changed the test to check the result independently of the result order.
Tests are green, so :+1:  from my side.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php [hunk 1]
```diff
@@ -937,8 +937,15 @@ public function findByProperties($term, $nodeTypeFilter, $workspace, $dimensions
         $queryBuilder = $this->createQueryBuilder($workspaces);
         $this->addDimensionJoinConstraintsToQueryBuilder($queryBuilder, $dimensions);
         $this->addNodeTypeFilterConstraintsToQueryBuilder($queryBuilder, $nodeTypeFilter);
-        // Convert to lowercase, then to json, and then trim quotes from json to have valid JSON escaping.
-        $likeParameter = '%' . trim(json_encode(UnicodeFunctions::strtolower($term), JSON_UNESCAPED_UNICODE), '"') . '%';
+
+        if (is_array($term)) {
```
Comment (hlubek) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Repository/NodeDataRepository.php]: This would only work for a single element array, right? Either we support multiple keys by iterating over each key and add multiple (ANDed) LIKE constraints or we throw an exception here, so the caller know exactly that this is not supported yet (instead of just finding no results or just by chance of ordering in the JSON blob - shudder).

**Comment by daniellienert** [comment]
Thanks @hlubek for the hint. I added an exception when someone tries to use an array with more than one key => value pair. The support for multiple values can be added in another PR. 
So it's once again mergeable. Or should we keep this #1 PR for historical reasons? :smile:

**Comment by aertmann** [comment]
good work @daniellienert, lets get this in.. ping @hlubek @kdambekalns @kitsunet

**Comment by aertmann** [comment]
can't have a failed first pull request 😄
